pageserver/src/remote_storage/rust_s3.rs

```diff
@@ -148,17 +148,49 @@ impl RemoteStorage for S3 {
             .with_context(|| format!("Failed to download s3 object with key {}", from.key()))?;
         if code != 200 {
             Err(anyhow::format_err!(
-                "Received non-200 exit code during downloading object from directory, code: {}",
+                "Received non-200 exit code during downloading object, code: {}",
                 code
             ))
         } else {
+            // we don't have to write vector into the destination this way, `to_write_all` would be enough.
+            // but we want to prepare for migration on `rusoto`, that has a streaming HTTP body instead here, with
+            // which it makes more sense to use `io::copy`.
             io::copy(&mut io::BufReader::new(data.as_slice()), to)
```

Comment:
I get wanting to use a streaming API, but should we still use BufReader? Can the slice alone be the input to io::copy

Reply:
The culprit here is `io::copy` signature:

> `pub async fn copy<'a, R, W>(reader: &'a mut R, writer: &'a mut W) -> io::Result<u64>`

`AsyncRead` is implemented for `&[u8]` only, not its mutable counterpart, so I have to wrap it.

I will wrap it in `std::io::Cursor` then, no need to use `BufReader`, but not sure what else besides those two could be used as `&mut impl AsyncRead`. 
Maybe there's a better method that just accepts non-mutable `AsyncRead` somewhere?

Reply:
What about
`io::copy(&mut data.as_slice(), ...)`

Reply:
Oh my, indeed. Thank you 😅